src/item/enchantment/EnchantmentHelper.php

```diff
@@ -0,0 +1,217 @@
+
+		for($x = -2; $x <= 2; $x++){
+			for($z = -2; $z <= 2; $z++){
+				// We only check blocks at a distance of 2 blocks from the enchanting table
+				if(abs($x) !== 2 && abs($z) !== 2){
+					continue;
+				}
+				for($y = 0; $y <= 1; $y++){
+					$block = $world->getBlock($tablePos->add($x, $y, $z));
+					if($block->getTypeId() !== BlockTypeIds::BOOKSHELF){
+						continue;
+					}
+
+					// Calculate the coordinates of the space between the bookshelf and the enchanting table
+					$spaceX = max(min($x, 1), -1);
+					$spaceZ = max(min($z, 1), -1);
+
+					$lowerSpaceBlock = $world->getBlock($tablePos->add($spaceX, 0, $spaceZ));
+					if($lowerSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+					$upperSpaceBlock = $world->getBlock($tablePos->add($spaceX, 1, $spaceZ));
+					if($upperSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+
+					$bookshelfCount++;
+					if($bookshelfCount === self::MAX_BOOKSHELF_COUNT){
+						return $bookshelfCount;
+					}
+				}
+			}
+		}
+
+		return $bookshelfCount;
+	}
+
+	private static function createEnchantOption(Random $random, Item $inputItem, int $optionCost, int $slot) : EnchantmentOption{
+		$cost = $optionCost;
+
+		$enchantability = self::getEnchantability($inputItem);
+		$cost = $cost + $random->nextRange(0, $enchantability >> 2) + $random->nextRange(0, $enchantability >> 2) + 1;
+		// Random bonus for enchantment cost between 0.85 and 1.15
+		$bonus = 1 + ($random->nextFloat() + $random->nextFloat() - 1) * 0.15;
+		$cost = (int) round($cost * $bonus);
+
+		$resultEnchantments = [];
+		$availableEnchantments = self::getAvailableEnchantments($cost, $inputItem);
+
+		$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+		if($lastEnchantment !== null){
+			$resultEnchantments[] = $lastEnchantment;
+
+			// With probability (cost + 1) / 50, continue adding enchantments
+			while($random->nextFloat() <= ($cost + 1) / 50){
+				// Remove from the list of available enchantments anything that conflicts
+				// with previously-chosen enchantments
+				$availableEnchantments = array_filter(
+					$availableEnchantments,
+					function(EnchantmentInstance $e) use ($lastEnchantment){
+						return $e->getType() !== $lastEnchantment->getType() &&
+							$e->getType()->isCompatibleWith($lastEnchantment->getType());
+					}
+				);
+
+				$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+				if($lastEnchantment === null){
+					break;
+				}
+
+				$resultEnchantments[] = $lastEnchantment;
+				$cost >>= 1;
+			}
+		}
+
+		return new EnchantmentOption($optionCost, $slot, $resultEnchantments, self::getRandomOptionName($random));
+	}
+
+	private static function getEnchantability(Item $item) : int{
+		if($item instanceof TieredTool){
+			return $item->getTier()->getEnchantability();
+		}
+		if($item instanceof Armor){
+			return $item->getMaterial()->getEnchantability();
+		}
+		return 1;
+	}
```

Comment:
But we'll have to add the $enchantability parameter to the constructor of each item that overwrites the parent constructor. And we will have to duplicate the enchantability value for each tool and armor from the same material. Is that okay? Maybe keep vanilla's idea that only tools and armor have different enchantability values, and other items have enchantability equal to 1?

Reply:
Yes, you won't be able to customize the enchantability value for custom tools and armor because you can't dynamically register your ToolTier value and ArmorMaterial, but the same goes for maxDurability, baseAttackPoints, baseEfficiency, etc. That's a different problem not related to enchantments, isn't it?

Reply:
Basically, I could leave $enchantability in ToolTier and ArmorMaterial, and just add in Item `getEnchantability` that accesses ToolTier / Material or just returns 1. Maybe that's what you meant? Then I didn't understand right away))